charms/istio-pilot/tests/unit/test_charm.py

```diff
@@ -336,6 +335,57 @@ def test_get_ingress_data_waiting_on_version(self, harness):
 
         assert "versions not found" in err.value.msg
 
+    @pytest.mark.parametrize(
+        "related_applications",
+        [
+            ([]),  # No related applications
+            (["other1"]),  # A single related application
+            (["other1", "other2", "other3"]),  # Multiple related applications
+        ],
+    )
+    def test_handle_istio_pilot_relation(self, related_applications, harness):
+        """Tests that the handle_istio_pilot_relation helper works as expected."""
+        # Assert
+        # Must be leader because we write to the application part of the relation data
+        model_name = "some-model"
+        expected_data = {
+            "service-name": f"istiod.{model_name}.svc",
+            "service-port": "15012",
+        }
+
+        harness.set_leader(True)
+        harness.set_model_name(model_name)
+
+        relation_info = [
+            add_istio_pilot_to_harness(harness, other_app=name) for name in related_applications
+        ]
+        harness.begin()
```

Comment:
quick check, should not we harness begin before adding units?

I am 90% sure it should be before

Reply:
the correct way depends on what we want to test. the data is stored into the relation correctly in either order, but the charm reacts differently depending on the order:
1) begin, add relations, add data: when you add relations, relation created and joined events will fire. then same when you add data (relation changed event). 
2) add relations, add data, begin: no relation events will fire, and your charm will begin with data already available

These tests go with (2) because under test is just the helper, not the whole charm logic. If we do (1), we are testing that:
- we observed the relation event
- everything in the charm works successfully up until we get to what we are testing
- the method under test works however the charm wants to use it
- the rest of the charm also works

(1) is more an integration test, whereas (2) is more a unit test. If we try to use (1) here, we will have to mock a ton of stuff out to get the same result